File: swordOffer/dp/translateNum.cc
#include <iostream>
#include <vector>

using namespace std;
class Solution {
 public:
  int translateNum(int num) {
    if (num < 10) {
      return 1;
    }

    int res = 1;
    int prev = 26;
    int q = 1;
    int s = 1;
    while (num > 0) {
      int m = num % 10;
      //   cout << "x: "<< m * 10 + prev << endl;
      if (int x = m * 10 + prev; x <= 25 && m != 0 && x != 0) {
        res = q + s;
      } else {
        res = q;
      }
      prev = m;
      s = q;
      q = res;
      num /= 10;
    }

    return res;
  }
};

File: swordOffer/go/dp/translateNum.go
package main

import "strconv"

func translateNum(num int) int {
	s := strconv.Itoa(num)
	if len(s) == 1 {
		return 1
	}

	p, q, r := 0, 0, 1
	for i := 0; i < len(s); i++ {
		p, q, r = q, r, 0
		r += q
		if i == 0 {
			continue
		}
		if x := s[i-1 : i+1]; x >= "10" && x <= "25" {
			r += p
		}
	}

	return r
}
